### ISSUE TITLE
Support for wrapfigure

### ISSUE THREAD

**Issue by kuon** [quarto-dev/quarto-cli#376]
I have some use case where I need the figure to be wrapped within the text.

I can achieve this with:

```
\begin{wrapfigure}{R}{6cm}
\centering
\includegraphics[width=5.5cm]{a.jpg}
\end{wrapfigure}
```

but it would be very handy to have it natively supported for both HTML and pdf output.

HTML could simple use `float` css property.

**Comment by dragonstyle** [comment]
This is a great suggestion and something that we'll plan on supporting in the future. Thanks!

**Comment by jmbuhr** [comment]
Just for reference and in case you want to take this even further: With html / css text can also wrap text around the outlines of a transparent png, which I used in this quarto-made scientific poster: https://hits-mbm-dev.github.io/paper-talin-loop/poster.html

![image](https://user-images.githubusercontent.com/17450586/156551140-a8032397-a282-463c-bbc3-0bfa39b70d42.png)

 (css: https://github.com/hits-mbm-dev/paper-talin-loop/blob/a4f94a20e1e88d72585da36727a3c424de673a3d/poster.css#L23-L33)

**Comment by jmbuhr** [comment]
No, this is a feature of html and css. There may be a LaTeX package that does something similar for pdf but I am no export on TeX.

**Comment by srvanderplas** [comment]
Any progress on this? It would be very useful to be able to wrap figures in PDF files for e.g. grant applications and other reports where pages are limited.

**Comment by mcanouil** [comment]
An open issue means it's still planed.
Updates are made in the issues.
If you don't see anything, then that means there was no updates to provide.

**Comment by JonasLoos** [comment]
As this is the first thing that shows up when searching for "quarto wrapfigure", I thought it might be helpful to share a workaround to get code block plots to be behave as a **wrapfigure for Latex PDF** output. I wrote a small lua filter to temporarily redefine the figure environment when the `wrapfigure` property is defined in the code block properties.

<details><summary>Click to see code and example</summary>

## Lua filter:

```lua
if FORMAT:match 'latex' then
  function Div(elem)
    -- Check if this div has wrapfigure attribute
    if elem.attributes and elem.attributes["wrapfigure"] then
      local wrap_pos, wrap_width = elem.attributes["wrapfigure"]:match("^%s*(%S+)%s+(%S+)%s*$")
      if wrap_pos == nil or wrap_width == nil then
        io.stderr:write("Error: wrapfigure attribute must be of the form 'pos width', e.g. 'R 0.5'. Found: '" .. elem.attributes["wrapfigure"] .. "'\n")
        return elem
      end
      return {
          pandoc.RawInline('latex', '\\let\\oldfigure\\figure\n\\let\\endoldfigure\\endfigure\n\\renewenvironment{figure}{}{}\n\\begin{wrapfigure}{' .. wrap_pos.. '}{' .. wrap_width .. '\\textwidth}\\centering\\captionsetup{format=plain, labelformat=simple}'),
          elem,
          pandoc.RawInline('latex', '\\end{wrapfigure}\\let\\figure\\oldfigure\n\\let\\endfigure\\endoldfigure')
      }
    end
    return elem
  end
end
```

Save this somewhere (e.g. under `assets/wrapfigure.lua`) and reference it in the file metadata or `_quarto.yml` like this:

```
filters:
  - assets/wrapfigure.lua
```


## Usage:

    ```{python}
    #| label: fig-test123
    #| fig-cap: Some caption
    #| wrapfigure: R 0.5
    
    import matplotlib.pyplot as plt
    plt.plot([1,2,3])
    plt.show()
    ```

Here `R` is the position (`R` -> right, `L` -> left), and `0.5` is the width, relative to `\textwidth`.

</details>

**Comment by cscheid** [comment]
That's a good workaround, thanks for sharing!

I'll note here that Quarto does support `fig-env` as an attribute to control the actual environment used for the figure. With that said, I unfortunately don't think that this would work with `wrapfigure`, because of the different required parameters.

@gordonwoodhull Do you think we could add something like https://sitandr.github.io/typst-examples-book/book/packages/wrapping.html to our Typst renderer?

I would be happy with shipping native `wrapfigure` support in Quarto if we had HTML, LaTeX and Typst support.

- HTML: https://getbootstrap.com/docs/5.0/utilities/float/
- LaTeX: `wrapfig.sty`/`wrapfigure`
- Typst: ...?

**Comment by gordonwoodhull** [comment]
That’s a good idea. 

They mentioned on the community call that Typst is adding built-in support for wrapping text around images soon. 

So if we add the dependency we should also plan to remove it soon.

**Comment by cscheid** [comment]
> Typst is adding built-in support for wrapping text around images soon.

Did you get a sense of what "soon" meant? If it's 0.13-ish, then we probably want to wait for them. If not, then we should do what you suggested instead.

**Comment by gordonwoodhull** [comment]
They were using words that sounded like “next release” but I wouldn’t want to bank on it because I know how software works. 

I think it involves some more ambitious layouts than just this feature.

**Comment by multimeric** [comment]
I've been working on solving this problem with a Quarto plugin: [multimeric/fwrappe](https://github.com/multimeric/fwrappe#readme). Currently only HTML and PDF/LaTeX are supported but I would welcome a typst implementation.

I do ultimately think this logic should live in quarto core because it's so fundamental, so I'm happy to contribute my code and be involved in discussions relating to that. [I've written some technical docs that might be helpful for such an implementation](https://multimeric.github.io/Fwrappe/implementation.html). It's quite a tricky problem in LaTeX that won't have an easy solution. Mine is just scratching the surface.

**Comment by cscheid** [comment]
@multimeric I agree, and we'd love to take your contribution! Do you want to start a PR? How can we help?

**Comment by multimeric** [comment]
I assume this doesn't make sense as a Quarto feature until all the output formats are supported (correct me if this is not true), which means it can't be done incrementally. This is why I did it as an extension. On the other hand, extensions aren't really "powerful" enough to do this feature well, because they don't have access to all the lifecycle hooks (see https://github.com/quarto-dev/quarto-cli/discussions/11749 and https://github.com/quarto-dev/quarto-cli/issues/11750), and also they don't have access to the much more extensive internal lua API. So I'm not sure the best way to proceed.

**Comment by cscheid** [comment]
> I assume this doesn't make sense as a Quarto feature until all the output formats are supported (correct me if this is not true), which means it can't be done incrementally.

We have a number of features that don't span across all formats. We prefer many- or all-format features, but formats are different from each other for good reasons, and so we don't hold this too strongly. In addition, the rendering of a wrapfigure can degrade gracefully into a regular figure + warning.

I think HTML + LaTeX is a good minimal version that I'd be happy to merge. If we supported Typst that would be even nicer; there's some [public-domain code to start from](https://github.com/ntjess/wrap-it)).
